internal/worker/sshsession/worker.go

```diff
@@ -0,0 +1,349 @@
+	// WatchSSHConnRequest creates a watcher and returns its ID for watching changes.
+	WatchSSHConnRequest(machineId string) (watcher.StringsWatcher, error)
+
+	// GetSSHConnRequest returns a ssh connection request by its connection request ID.
+	GetSSHConnRequest(arg string) (params.SSHConnRequest, error)
+}
+
+// WorkerConfig encapsulates the configuration options for
+// instantiating a new lease ssh session worker.
+type WorkerConfig struct {
+	Logger           Logger
+	MachineId        string
+	FacadeClient     FacadeClient
+	ConnectionGetter ConnectionGetter
+}
+
+// Validate checks whether the worker configuration settings are valid.
+func (cfg WorkerConfig) Validate() error {
+	if cfg.Logger == nil {
+		return errors.NotValidf("nil Logger")
+	}
+	if cfg.MachineId == "" {
+		return errors.NotValidf("empty MachineId")
+	}
+	if cfg.FacadeClient == nil {
+		return errors.NotValidf("nil FacadeClient")
+	}
+	if cfg.ConnectionGetter == nil {
+		return errors.NotValidf("nil ConnectionGetter")
+	}
+
+	return nil
+}
+
+// sshSessionWorker is a worker that enables SSH connections to a machine.
+// It watches for incoming SSH connections to the controller's SSH server over
+// a facade, and when notified initiates a connection to the controller's SSH server.
+// When the controller's SSH server receives this connection, it initiates its own
+// SSH connection down the tunnel and the sshSessionWorker pipes it to the SSHD
+// on the machine.
+type sshSessionWorker struct {
+	catacomb catacomb.Catacomb
+
+	logger           Logger
+	machineId        string
+	facadeClient     FacadeClient
+	connectionGetter ConnectionGetter
+}
+
+// NewWorker returns an SSH session worker.
+func NewWorker(cfg WorkerConfig) (worker.Worker, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	w := &sshSessionWorker{
+		logger:           cfg.Logger,
+		machineId:        cfg.MachineId,
+		facadeClient:     cfg.FacadeClient,
+		connectionGetter: cfg.ConnectionGetter,
+	}
+
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+	})
+
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return w, nil
+}
+
+// loop starts the workers main loop. It watches for incoming SSH connection requests
+// and handles them by initiating a reverse SSH connection to the controller and piping
+// it to the local sshd of the machine agent's machine.
+func (w *sshSessionWorker) loop() error {
+	connRequestWatcher, err := w.facadeClient.WatchSSHConnRequest(w.machineId)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if err := w.catacomb.Add(connRequestWatcher); err != nil {
+		return errors.Trace(err)
+	}
+
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case changes := <-connRequestWatcher.Changes():
+			for _, connId := range changes {
+				err := w.handleConnection(connId)
+				if err != nil {
+					w.logger.Errorf("failed to handle connection %q: %v", connId, err)
```

Comment:
You're logging at every layer, so your log will be duplicated 3+ times.
If you're returning an error, you're handling it, so don't log if you return the error. Only log it once here.

Reply:
Unless I wrap the errors it might be difficult to know the exact lines when debugging as the errors emitted could be from the underlying package. Remember, all we have here are logs nothing else and for debugging they're super useful.